File: AdvantageFramework/Database/Procedures/Alert.vb
Namespace Database.Procedures.Alert

    <HideModuleName()> _
    Public Module Methods

#Region " Constants "



#End Region

#Region " Enum "



#End Region

#Region " Variables "



#End Region

#Region " Properties "



#End Region

#Region " Methods "

        Public Function LoadTaskWorkItem(ByVal DbContext As AdvantageFramework.Database.DbContext,
                                                         ByVal JobNumber As Integer, ByVal JobComponentNumber As Short, ByVal TaskSequenceNumber As Short) As AdvantageFramework.Database.Entities.Alert

            LoadTaskWorkItem = (From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                                Where (Alert.AlertLevel = "BRD" Or Alert.AlertCategoryID = 71) And
                                       Alert.JobNumber = JobNumber And
                                       Alert.JobComponentNumber = JobComponentNumber And
                                       Alert.TaskSequenceNumber = TaskSequenceNumber
                                Select Alert).SingleOrDefault

        End Function
        Public Function LoadAssignmentsByJobAndComponent(ByVal DbContext As AdvantageFramework.Database.DbContext,
                                                         ByVal JobNumber As Integer, ByVal JobComponentNumber As Short) As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Alert)

            LoadAssignmentsByJobAndComponent = From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                                               Where Alert.JobNumber = JobNumber And Alert.JobComponentNumber = JobComponentNumber And
                                                     Alert.AlertAssignmentTemplateID IsNot Nothing And Alert.AlertStateID IsNot Nothing
                                               Select Alert

        End Function

        Public Function LoadByConceptShareReviewID(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal ConceptShareReviewID As Integer) As AdvantageFramework.Database.Entities.Alert

            Try

                LoadByConceptShareReviewID = (From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                                              Where Alert.ConceptShareReviewID = ConceptShareReviewID
                                              Select Alert).SingleOrDefault

            Catch ex As Exception
                LoadByConceptShareReviewID = Nothing
            End Try

        End Function
        Public Function LoadByAlertID(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal AlertID As Long) As AdvantageFramework.Database.Entities.Alert

            Try

                LoadByAlertID = (From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                                 Where Alert.ID = AlertID
                                 Select Alert).SingleOrDefault

            Catch ex As Exception
                LoadByAlertID = Nothing
            End Try

        End Function
        Public Function Load(ByVal DbContext As AdvantageFramework.Database.DbContext) As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Alert)

            Load = From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                   Select Alert

        End Function
        Public Function LoadByTypeIDAndClientAndDivisionAndProductCode(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal TypeID As Integer, ByVal ClientCode As String, ByVal DivisionCode As String, ByVal ProductCode As String) As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Alert)

            LoadByTypeIDAndClientAndDivisionAndProductCode = From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                                                             Where Alert.AlertTypeID = TypeID AndAlso
                                                                   Alert.ClientCode = ClientCode AndAlso
                                                                   Alert.DivisionCode = DivisionCode AndAlso
                                                                   Alert.ProductCode = ProductCode
                                                             Select Alert

        End Function
        Public Function LoadByNonTaskID(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal NonTaskID As Long) As AdvantageFramework.Database.Entities.Alert

            Try

                LoadByNonTaskID = (From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                                   Where Alert.NonTaskID = NonTaskID
                                   Select Alert).SingleOrDefault

            Catch ex As Exception
                LoadByNonTaskID = Nothing
            End Try

        End Function
        Public Function LoadByJobVersionID(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal JobVersionID As Long) As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Alert)

            Try

                LoadByJobVersionID = From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                                     Where Alert.JobVersionID = JobVersionID
                                     Select Alert

            Catch ex As Exception
                LoadByJobVersionID = Nothing
            End Try

        End Function
        Public Function LoadClientContactAlerts(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal ClientContactID As Integer, ByVal Skip As Integer, ByVal Take As Integer) As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Alert)

            'LoadClientContactAlerts = LoadEmployeeAlertsAndAssignments(DbContext, EmployeeCode).Where(Function(Alert) Alert.AlertRecipients.Any(Function(AlertRecipient) AlertRecipient.IsCurrentNotify Is Nothing OrElse AlertRecipient.IsCurrentNotify = 0))
            LoadClientContactAlerts = (From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                                       Join AlertRecipients In ClientPortalAlertRecipient.Load(DbContext)
                                       On Alert.ID Equals AlertRecipients.AlertID
                                       Where AlertRecipients.ClientContactID = ClientContactID
                                       Order By Alert.LastUpdated Descending
                                       Select Alert).Skip(Skip).Take(Take)

        End Function
        Public Function LoadEmployeeAlerts(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal EmployeeCode As String, ByVal Skip As Integer, ByVal Take As Integer) As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Alert)

            'LoadEmployeeAlerts = LoadEmployeeAlertsAndAssignments(DbContext, EmployeeCode).Where(Function(Alert) Alert.AlertRecipients.Any(Function(AlertRecipient) AlertRecipient.IsCurrentNotify Is Nothing OrElse AlertRecipient.IsCurrentNotify = 0))
            LoadEmployeeAlerts = (From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                                  Join AlertRecipients In AlertRecipient.Load(DbContext)
                                  On Alert.ID Equals AlertRecipients.AlertID
                                  Where AlertRecipients.EmployeeCode = EmployeeCode _
                                  AndAlso (AlertRecipients.IsCurrentNotify Is Nothing OrElse AlertRecipients.IsCurrentNotify = 0)
                                  Order By Alert.LastUpdated Descending
                                  Select Alert).Skip(Skip).Take(Take)

        End Function
        Public Function LoadEmployeeAssignments(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal EmployeeCode As String, ByVal Skip As Integer, ByVal Take As Integer) As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Alert)

            'LoadEmployeeAssignments = LoadEmployeeAlertsAndAssignments(DbContext, EmployeeCode).Where(Function(Alert) Alert.AlertRecipients.Any(Function(AlertRecipient) AlertRecipient.IsCurrentNotify = 1))
            LoadEmployeeAssignments = (From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                                       Join AlertRecipients In AlertRecipient.Load(DbContext)
                                       On Alert.ID Equals AlertRecipients.AlertID
                                       Where AlertRecipients.EmployeeCode = EmployeeCode _
                                       And AlertRecipients.IsCurrentNotify = 1
                                       Order By Alert.LastUpdated Descending
                                       Select Alert).Skip(Skip).Take(Take)

        End Function
        Public Function LoadEmployeeAlertsAndAssignments(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal EmployeeCode As String, ByVal Skip As Integer, ByVal Take As Integer) As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Alert)

            Dim query As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Alert)
            Dim list As List(Of Integer)

            list = (From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                    Join AlertRecipients In AlertRecipient.Load(DbContext)
                    On Alert.ID Equals AlertRecipients.AlertID
                    Where AlertRecipients.EmployeeCode = EmployeeCode
                    Order By Alert.LastUpdated Descending
                    Select Alert.ID).Distinct().ToList()

            query = (From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                     Where list.Contains(Alert.ID)
                     Order By Alert.ID
                     Select Alert)

            Return query.Skip(Skip).Take(Take)

        End Function
        Public Function LoadDigitalAssetReviewsByJobAndComponentNumber(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal JobNumber As Integer, ByVal JobComponentNumber As Short) As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Alert)

            LoadDigitalAssetReviewsByJobAndComponentNumber = From Alert In DbContext.GetQuery(Of Database.Entities.Alert)
                                                             Where Alert.JobNumber = JobNumber AndAlso
                                                                   Alert.JobComponentNumber = JobComponentNumber AndAlso
                                                                   Alert.AlertTypeID = 15
                                                             Select Alert

        End Function

        Public Function Insert(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal AlertTypeID As Integer, ByVal AlertCategoryID As Integer, _
                               ByVal Subject As String, ByVal Body As String, _
                               ByVal GeneratedDate As Nullable(Of Date), ByVal PriorityLevel As Nullable(Of Short), _
                               ByVal EmployeeCode As String, ByVal UserCode As String, _
                               ByVal AlertLevel As String, ByVal EmailBody As String, _
                               ByRef Alert As AdvantageFramework.Database.Entities.Alert) As Boolean

            'objects
            Dim Inserted As Boolean = False

            Try

                Alert = New AdvantageFramework.Database.Entities.Alert

                Alert.DbContext = DbContext
                Alert.AlertTypeID = AlertTypeID
                Alert.AlertCategoryID = AlertCategoryID
                Alert.Subject = Subject
                Alert.Body = Body
                Alert.GeneratedDate = GeneratedDate
                Alert.PriorityLevel = PriorityLevel
                Alert.EmployeeCode = EmployeeCode
                Alert.UserCode = UserCode
                Alert.AlertLevel = AlertLevel
                Alert.EmailBody = EmailBody
                Alert.LastUpdated = GeneratedDate

                Inserted = Insert(DbContext, Alert)

            Catch ex As Exception
                Inserted = False
            Finally
                Insert = Inserted
            End Try

        End Function
        Public Function Insert(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal Alert As AdvantageFramework.Database.Entities.Alert) As Boolean

            'objects
            Dim Inserted As Boolean = False
            Dim IsValid As Boolean = True
            Dim ErrorText As String = ""

            Try

                DbContext.Alerts.Add(Alert)

                ErrorText = Alert.ValidateEntity(IsValid)

                If IsValid Then

                    Try

                        Alert.ID = (From Entity In AdvantageFramework.Database.Procedures.Alert.Load(DbContext) _
                                     Select Entity.ID).Max + 1

                    Catch ex As Exception
                        Alert.ID = 1
                    End Try

                    DbContext.SaveChanges()

                    Inserted = True

                Else

                    AdvantageFramework.Navigation.ShowMessageBox(ErrorText)

                End If

            Catch ex As Exception
                Inserted = False
            Finally
                Insert = Inserted
            End Try

        End Function
        Public Function Update(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal Alert As AdvantageFramework.Database.Entities.Alert) As Boolean

            'objects
            Dim Updated As Boolean = False
            Dim IsValid As Boolean = True
            Dim ErrorText As String = ""

            Try

                DbContext.UpdateObject(Alert)

                ErrorText = Alert.ValidateEntity(IsValid)

                If IsValid Then

                    DbContext.SaveChanges()

                    Updated = True

                Else

                    AdvantageFramework.Navigation.ShowMessageBox(ErrorText)

                End If

            Catch ex As Exception
                Updated = False
            Finally
                Update = Updated
            End Try

        End Function
        Public Function Delete(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal Alert As AdvantageFramework.Database.Entities.Alert) As Boolean

            'objects
            Dim Deleted As Boolean = True
            Dim IsValid As Boolean = True
            Dim ErrorText As String = ""

            Try

                If IsValid Then

                    DbContext.DeleteEntityObject(Alert)

                    DbContext.SaveChanges()

                    Deleted = True

                Else

                    AdvantageFramework.Navigation.ShowMessageBox(ErrorText)

                End If

            Catch ex As Exception
                Deleted = False
            Finally
                Delete = Deleted
            End Try

        End Function

#End Region

    End Module

End Namespace

File: AdvantageFramework.Core/Database/Procedures/Alert.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AdvantageFramework.Core.Database.Procedures
{
    public static partial class Alert
    {
        #region  Constants 



        #endregion

        #region  Enum 



        #endregion

        #region  Variables 



        #endregion

        #region  Properties 



        #endregion

        #region  Methods 

        public static AdvantageFramework.Core.Database.Entities.Alert LoadTaskWorkItem(AdvantageFramework.Core.Database.DbContext DbContext, int JobNumber, short JobComponentNumber, short TaskSequenceNumber)
        {
            AdvantageFramework.Core.Database.Entities.Alert LoadTaskWorkItemRet = default;
            LoadTaskWorkItemRet = (from Alert in DbContext.Alerts.AsQueryable()
                                   where (Alert.AlertLevel == "BRD" | Alert.AlertCatId == 71) & Alert.JobNumber == JobNumber & Alert.JobComponentNbr == JobComponentNumber & Alert.TaskSeqNbr == TaskSequenceNumber
                                   select Alert).SingleOrDefault();
            return LoadTaskWorkItemRet;
        }

        public static System.Linq.IQueryable<Entities.Alert> LoadAssignmentsByJobAndComponent(AdvantageFramework.Core.Database.DbContext DbContext, int JobNumber, short JobComponentNumber)
        {
            System.Linq.IQueryable<Entities.Alert> LoadAssignmentsByJobAndComponentRet = default;
            LoadAssignmentsByJobAndComponentRet = from Alert in DbContext.Alerts.AsQueryable()
                                                  where Alert.JobNumber == JobNumber & Alert.JobComponentNbr == JobComponentNumber & Alert.AlrtNotifyHdrId is object & Alert.AlertStateId is object
                                                  select Alert;
            return LoadAssignmentsByJobAndComponentRet;
        }

        public static AdvantageFramework.Core.Database.Entities.Alert LoadByConceptShareReviewID(AdvantageFramework.Core.Database.DbContext DbContext, int ConceptShareReviewID)
        {
            AdvantageFramework.Core.Database.Entities.Alert LoadByConceptShareReviewIDRet = default;
            try
            {
                LoadByConceptShareReviewIDRet = (from Alert in DbContext.Alerts.AsQueryable()
                                                 where Alert.CsReviewId == ConceptShareReviewID
                                                 select Alert).SingleOrDefault();
            }
            catch (Exception )
            {
                LoadByConceptShareReviewIDRet = default;
            }

            return LoadByConceptShareReviewIDRet;
        }

        public static AdvantageFramework.Core.Database.Entities.Alert LoadByAlertID(AdvantageFramework.Core.Database.DbContext DbContext, long AlertID)
        {
            AdvantageFramework.Core.Database.Entities.Alert LoadByAlertIDRet = default;
            try
            {
                LoadByAlertIDRet = (from Alert in DbContext.Alerts.AsQueryable()
                                    where Alert.AlertId == AlertID
                                    select Alert).SingleOrDefault();
            }
            catch (Exception )
            {
                LoadByAlertIDRet = default;
            }

            return LoadByAlertIDRet;
        }

        public static System.Linq.IQueryable<Entities.Alert> Load(AdvantageFramework.Core.Database.DbContext DbContext)
        {
            System.Linq.IQueryable<Entities.Alert> LoadRet = default;
            LoadRet = from Alert in DbContext.Alerts.AsQueryable()
                      select Alert;
            return LoadRet;
        }

        public static System.Linq.IQueryable<Entities.Alert> LoadByTypeIDAndClientAndDivisionAndProductCode(AdvantageFramework.Core.Database.DbContext DbContext, int TypeID, string ClientCode, string DivisionCode, string ProductCode)
        {
            System.Linq.IQueryable<Entities.Alert> LoadByTypeIDAndClientAndDivisionAndProductCodeRet = default;
            LoadByTypeIDAndClientAndDivisionAndProductCodeRet = from Alert in DbContext.Alerts.AsQueryable()
                                                                where Alert.AlertTypeId == TypeID && Alert.ClCode == ClientCode && Alert.DivCode == DivisionCode && Alert.PrdCode == ProductCode
                                                                select Alert;
            return LoadByTypeIDAndClientAndDivisionAndProductCodeRet;
        }

        public static AdvantageFramework.Core.Database.Entities.Alert LoadByNonTaskID(AdvantageFramework.Core.Database.DbContext DbContext, long NonTaskID)
        {
            AdvantageFramework.Core.Database.Entities.Alert LoadByNonTaskIDRet = default;
            try
            {
                LoadByNonTaskIDRet = (from Alert in DbContext.Alerts.AsQueryable()
                                      where Alert.NonTaskId == NonTaskID
                                      select Alert).SingleOrDefault();
            }
            catch (Exception )
            {
                LoadByNonTaskIDRet = default;
            }

            return LoadByNonTaskIDRet;
        }

        public static System.Linq.IQueryable<Entities.Alert> LoadByJobVersionID(AdvantageFramework.Core.Database.DbContext DbContext, long JobVersionID)
        {
            System.Linq.IQueryable<Entities.Alert> LoadByJobVersionIDRet = default;
            try
            {
                LoadByJobVersionIDRet = from Alert in DbContext.Alerts.AsQueryable()
                                        where Alert.JobVerHdrId == JobVersionID
                                        select Alert;
            }
            catch (Exception )
            {
                LoadByJobVersionIDRet = default;
            }

            return LoadByJobVersionIDRet;
        }

        public static System.Linq.IQueryable<Entities.Alert> LoadClientContactAlerts(AdvantageFramework.Core.Database.DbContext DbContext, int ClientContactID, int Skip, int Take)
        {
            System.Linq.IQueryable<Entities.Alert> LoadClientContactAlertsRet = default;

            // LoadClientContactAlerts = LoadEmployeeAlertsAndAssignments(DbContext, EmployeeCode).Where(Function(Alert) Alert.AlertRecipients.Any(Function(AlertRecipient) AlertRecipient.IsCurrentNotify Is Nothing OrElse AlertRecipient.IsCurrentNotify = 0))
            LoadClientContactAlertsRet = (from Alert in DbContext.Alerts.AsQueryable()
                                          join AlertRecipients in ClientPortalAlertRecipient.Load(DbContext) on Alert.AlertId equals AlertRecipients.AlertId
                                          where AlertRecipients.CdpContactId == ClientContactID
                                          orderby Alert.LastUpdated descending
                                          select Alert).Skip(Skip).Take(Take);
            return LoadClientContactAlertsRet;
        }

        public static System.Linq.IQueryable<Entities.Alert> LoadEmployeeAlerts(AdvantageFramework.Core.Database.DbContext DbContext, string EmployeeCode, int Skip, int Take)
        {
            System.Linq.IQueryable<Entities.Alert> LoadEmployeeAlertsRet = default;

            // LoadEmployeeAlerts = LoadEmployeeAlertsAndAssignments(DbContext, EmployeeCode).Where(Function(Alert) Alert.AlertRecipients.Any(Function(AlertRecipient) AlertRecipient.IsCurrentNotify Is Nothing OrElse AlertRecipient.IsCurrentNotify = 0))
            LoadEmployeeAlertsRet = (from Alert in DbContext.Alerts.AsQueryable()
                                     join AlertRecipients in AlertRecipient.Load(DbContext) on Alert.AlertId equals AlertRecipients.AlertId
                                     where AlertRecipients.EmpCode == EmployeeCode && (AlertRecipients.CurrentNotify == null || AlertRecipients.CurrentNotify == 0)
                                     orderby Alert.LastUpdated descending
                                     select Alert).Skip(Skip).Take(Take);
            return LoadEmployeeAlertsRet;
        }

        public static System.Linq.IQueryable<Entities.Alert> LoadEmployeeAssignments(AdvantageFramework.Core.Database.DbContext DbContext, string EmployeeCode, int Skip, int Take)
        {
            System.Linq.IQueryable<Entities.Alert> LoadEmployeeAssignmentsRet = default;

            // LoadEmployeeAssignments = LoadEmployeeAlertsAndAssignments(DbContext, EmployeeCode).Where(Function(Alert) Alert.AlertRecipients.Any(Function(AlertRecipient) AlertRecipient.IsCurrentNotify = 1))
            LoadEmployeeAssignmentsRet = (from Alert in DbContext.Alerts.AsQueryable()
                                          join AlertRecipients in AlertRecipient.Load(DbContext) on Alert.AlertId equals AlertRecipients.AlertId
                                          where AlertRecipients.EmpCode == EmployeeCode & AlertRecipients.CurrentNotify == 1
                                          orderby Alert.LastUpdated descending
                                          select Alert).Skip(Skip).Take(Take);
            return LoadEmployeeAssignmentsRet;
        }

        public static System.Linq.IQueryable<Entities.Alert> LoadEmployeeAlertsAndAssignments(AdvantageFramework.Core.Database.DbContext DbContext, string EmployeeCode, int Skip, int Take)
        {
            System.Linq.IQueryable<Entities.Alert> query;
            List<int> list;
            list = (from Alert in DbContext.Alerts.AsQueryable()
                    join AlertRecipients in AlertRecipient.Load(DbContext) on Alert.AlertId equals AlertRecipients.AlertId
                    where AlertRecipients.EmpCode == EmployeeCode
                    orderby Alert.LastUpdated descending
                    select Alert.AlertId).Distinct().ToList();
            query = (from Alert in DbContext.Alerts.AsQueryable()
                     where list.Contains(Alert.AlertId)
                     orderby Alert.AlertId
                     select Alert);
            return query.Skip(Skip).Take(Take);
        }

        public static System.Linq.IQueryable<Entities.Alert> LoadDigitalAssetReviewsByJobAndComponentNumber(AdvantageFramework.Core.Database.DbContext DbContext, int JobNumber, short JobComponentNumber)
        {
            System.Linq.IQueryable<Entities.Alert> LoadDigitalAssetReviewsByJobAndComponentNumberRet = default;
            LoadDigitalAssetReviewsByJobAndComponentNumberRet = from Alert in DbContext.Alerts.AsQueryable()
                                                                where Alert.JobNumber == JobNumber && Alert.JobComponentNbr == JobComponentNumber && Alert.AlertTypeId == 15
                                                                select Alert;
            return LoadDigitalAssetReviewsByJobAndComponentNumberRet;
        }

        //public static bool Insert(AdvantageFramework.Core.Database.DbContext DbContext, int AlertTypeID, int AlertCategoryID, string Subject, string Body, DateTime? GeneratedDate, short? PriorityLevel, string EmployeeCode, string UserCode, string AlertLevel, string EmailBody, ref AdvantageFramework.Core.Database.Entities.Alert Alert)




        //{
        //    bool InsertRet = default;

        //    // objects
        //    bool Inserted = false;
        //    try
        //    {
        //        Alert = new AdvantageFramework.Core.Database.Entities.Alert();
        //        Alert.DbContext = DbContext;
        //        Alert.AlertTypeID = AlertTypeID;
        //        Alert.AlertCategoryID = AlertCategoryID;
        //        Alert.Subject = Subject;
        //        Alert.Body = Body;
        //        Alert.GeneratedDate = GeneratedDate;
        //        Alert.PriorityLevel = PriorityLevel;
        //        Alert.EmployeeCode = EmployeeCode;
        //        Alert.UserCode = UserCode;
        //        Alert.AlertLevel = AlertLevel;
        //        Alert.EmailBody = EmailBody;
        //        Alert.LastUpdated = GeneratedDate;
        //        Inserted = Insert(DbContext, Alert);
        //    }
        //    catch (Exception ex)
        //    {
        //        Inserted = false;
        //    }
        //    finally
        //    {
        //        InsertRet = Inserted;
        //    }

        //    return InsertRet;
        //}

        //public static bool Insert(AdvantageFramework.Core.Database.DbContext DbContext, AdvantageFramework.Core.Database.Entities.Alert Alert)
        //{
        //    bool InsertRet = default;

        //    // objects
        //    bool Inserted = false;
        //    bool IsValid = true;
        //    string ErrorText = "";
        //    try
        //    {
        //        DbContext.Alerts.Add(Alert);
        //        //ErrorText = Alert.ValidateEntity(IsValid);
        //        if (IsValid)
        //        {
        //            try
        //            {
        //                Alert.AlertId = (from Entity in Load(DbContext)
        //                            select Entity.AlertId).Max() + 1;
        //            }
        //            catch (Exception )
        //            {
        //                Alert.AlertId = 1;
        //            }

        //            DbContext.SaveChanges();
        //            Inserted = true;
        //        }
        //        else
        //        {
        //            //AdvantageFramework.Navigation.ShowMessageBox(ErrorText);
        //        }
        //    }
        //    catch (Exception ex)
        //    {
        //        Inserted = false;
        //    }
        //    finally
        //    {
        //        InsertRet = Inserted;
        //    }

        //    return InsertRet;
        //}

        public static bool Update(AdvantageFramework.Core.Database.DbContext DbContext, AdvantageFramework.Core.Database.Entities.Alert Alert)
        {
            bool UpdateRet = default;

            // objects
            bool Updated = false;
            bool IsValid = true;
            string ErrorText = "";
            try
            {
                //TODO: Actually update the alert
                //DbContext.UpdateObject(Alert);
                //ErrorText = Alert.ValidateEntity(IsValid);
                if (IsValid)
                {
                    DbContext.SaveChanges();
                    Updated = true;
                }
                else
                {
                    //AdvantageFramework.Navigation.ShowMessageBox(ErrorText);
                }
            }
            catch (Exception )
            {
                Updated = false;
            }
            finally
            {
                UpdateRet = Updated;
            }

            return UpdateRet;
        }

        //public static bool Delete(AdvantageFramework.Core.Database.DbContext DbContext, AdvantageFramework.Core.Database.Entities.Alert Alert)
        //{
        //    bool DeleteRet = default;

        //    // objects
        //    bool Deleted = true;
        //    bool IsValid = true;
        //    string ErrorText = "";
        //    try
        //    {
        //        if (IsValid)
        //        {
        //            DbContext.DeleteEntityObject(Alert);
        //            DbContext.SaveChanges();
        //            Deleted = true;
        //        }
        //        else
        //        {
        //            //AdvantageFramework.Navigation.ShowMessageBox(ErrorText);
        //        }
        //    }
        //    catch (Exception )
        //    {
        //        Deleted = false;
        //    }
        //    finally
        //    {
        //        DeleteRet = Deleted;
        //    }

        //    return DeleteRet;
        //}

        #endregion
    }
}
